src/cuda_utils.h

```diff
@@ -118,6 +134,36 @@ Status GetSupportedGPUs(
 /// \return The error status. A non-OK status means the target GPU is
 /// not supported.
 Status SupportsIntegratedZeroCopy(const int gpu_id, bool* zero_copy_support);
+
+
+/// Set the CUDA context to the specified device ID
+/// It will rollback to the previous device upon destruction.
+class ScopedSetDevice {
+ public:
+  ScopedSetDevice(int device)
+  {
+    overriden_ = false;
+
+    prev_device_ = device;
+    cudaGetDevice(&prev_device_);
```

Comment:
question: is device process wide or thread local? if process wide is there anything we can / need to do for thread safety?

If one thread sets a device and another switches it is there any error handling we can do?

Reply:
It is per thread: https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g159587909ffa0791bbe4b40187a4c6bb